src/balancerd/src/lib.rs

```diff
@@ -244,10 +287,117 @@ impl BalancerService {
     }
 }
 
+#[allow(clippy::unused_async)]
+async fn handle_readiness_check() -> impl IntoResponse {
+    (StatusCode::OK, "ready")
+}
+
+struct InternalHttpServer {
+    router: Router,
+}
+
+impl mz_server_core::Server for InternalHttpServer {
+    const NAME: &'static str = "internal_http";
+
+    fn handle_connection(&self, conn: TcpStream) -> mz_server_core::ConnectionHandler {
+        let router = self.router.clone();
+        Box::pin(async {
+            let http = hyper::server::conn::Http::new();
+            http.serve_connection(conn, router).err_into().await
+        })
+    }
+}
+
+/// Wraps an IntGauge and automatically `inc`s on init and `drop`s on drop. Callers should not call
+/// `inc().`. Useful for handling multiple task exit points, for example in the case of a panic.
+struct GaugeGuard {
+    gauge: IntGauge,
+}
+
+impl From<IntGauge> for GaugeGuard {
+    fn from(gauge: IntGauge) -> Self {
+        let _self = Self { gauge };
+        _self.gauge.inc();
+        _self
+    }
+}
+
+impl Drop for GaugeGuard {
+    fn drop(&mut self) {
+        self.gauge.dec();
+    }
+}
+
+#[derive(Clone, Debug)]
+struct ServerMetricsConfig {
+    connection_status: IntCounterVec,
+    active_connections: IntGaugeVec,
+}
+
+impl ServerMetricsConfig {
+    fn register_into(registry: &MetricsRegistry) -> Self {
+        let connection_status = registry.register(metric!(
+            name: "mz_balancer_connection_status",
+            help: "Count of completed network connections, by status",
+            var_labels: ["source", "status"],
+        ));
+        let active_connections = registry.register(metric!(
+            name: "mz_balancer_connection_active",
```

Comment:
would recommend using two counters: # of connections ever established, and # of connections completed. then the active count can be inferred by calculating them, while giving a bit more information overall.

Reply:
not sure what `source` is, but do we have any information on what the destination environment is for each connection? could be useful to surface that (at the risk of increased cardinality, but today I think that'd be fine)